Prevent private functions from being overridden

Marenz opened the pull request:
fixes #8417

Marenz committed:
Disallow private functions from being overridden

Marenz commented:
the failed zepplin test seems to be because it uses `override` etc which this compiler version doesn't support. Not sure why it does that.. maybe the test files are fetched externally.

Don't know the problem with the solcjs test. Help appreciated.

leonardoalt commented:
@axic any idea what the solc-js problem could be here? Can it be related to the release version?

ekpyron commented:
We probably need to patch soljson.js here, let me check.

ekpyron commented:
So adding
```
# Patch soljson.js to provide backwards-compatibility with older emscripten versions
echo ";/* backwards compatibility */ Module['Runtime'] = Module;" >> $BUILD_DIR/libsolc/soljson.js
```
to ``build_emscripten.sh`` right before copying to the ``upload/`` directory should fix this, but thinking about it, it shouldn't be necessary and may be an issue in solc-js - I'll quickly check there.

ekpyron commented:
Hm, that's weird... the ``soljson.js`` built by CI here, *does* correctly contain that workaround...

ekpyron commented:
Ah, right, I know why this happens. This is (IMHO) pretty annoying about the ``solc-js`` tests... it will download the most recent version of soljson.js on its own (i.e. 0.6.4) and run against that as well without being asked... and in that we don't have the workarounds anymore...

ekpyron commented:
I.e. the failure is entirely harmless, but it will be hard to fix...

ekpyron commented:
One option would be to add
``sed -i -e 's/latest/v0.5.16+commit.9c3226ce/' test/package.js``
Right before running the tests in ``test/solcjsTests.sh``.
Or actually maybe better ``sed -i -e 's/latest/v0.5.0+commit.1d4f565a/' test/package.js``.

It's running on solc-js version 0.5.0 and we don't want to test that against latest (i.e. 0.6.4), since that is bound to break. Actually we don't want to test against any remote version here at all :-D... but the sanest choice, since we kind of have to, would be ``v0.5.0``.

ekpyron commented:
The other test run: ``t_ems_compile_ext_zeppelin``: if we want that to pass we need to downgrade to a version of zeppelin that is 0.5 compatible. I.e. replace ``master`` in
``test/externalTests/zeppelin.sh`` with some appropriate branch/tag/commit.

leonardoalt commented:
Sigh.. would be fine with not passing the zeppelin test

chriseth commented on Changelog.md:
```diff
@@ -1,3 +1,8 @@
+### 0.5.17 (unreleased)
+
+Bugfixes:
+ * Type Checker: Prevent overriding of private functions.
```
```suggestion
 * Type Checker: Disallow overriding of private functions.
```

Marenz commented:
Update:
* Added test for multiple inheritance
* Using version 2.5.0 of zepplin, which is before they updated to 0.6.0
* Added fix daniel suggested to solcjs test

chriseth commented on libsolidity/analysis/ContractLevelChecker.cpp:
```diff
@@ -203,6 +203,9 @@ void ContractLevelChecker::checkFunctionOverride(FunctionDefinition const& _func
 			stateMutabilityToString(_function.stateMutability()) +
 			"\"."
 		);
+	if (_function.visibility() == Declaration::Visibility::Private)
```
This actually checks "private functions cannot override". Can you add a test for different combinations of publc / private in the base an derived?

Marenz replied:
Done. All fail as they should :)

chriseth replied:
But the error messages are wrong: https://github.com/ethereum/solidity/pull/8455/files#diff-302ae33fa6301f5b2401b6e8cd02484a

You need to test for `_super.visibility()`

Marenz replied:
oh, true. Updated

chriseth commented:
Looks good! I'll make this the release PR directly.
